Frontend Routing: Always render standalone plugin pages using the `<AppRootPage>`

leventebalogh opened the pull request:
**Related issue:** https://github.com/grafana/cloud-onboarding/issues/2273

**Related PRs**
- [Backend: add pluginId to standalone plugin page NavLinks](https://github.com/grafana/grafana/pull/57769)
- [Connections: support standalone plugin pages](https://github.com/grafana/grafana/pull/57772)

**Registering plugin pages under core URLs only work if those are set explicitly in the INI settings.** Without these settings the `isCorePage=true` pages have no effect, they are "just exposed". (Please see the examples below for how to use it).

### Why?
With standalone plugin pages (`[navigation.app_standalone_pages]`) we allow server admins to move single plugin pages under a different section in the navtree. This can be even more powerful if plugins can extend / override core functionality by rendering under URLs other than `/a/...`. (This is something that we explicitly needed for the Connections project as well, as certain core functionality needed to be extended if we are running in the Cloud).

### What changed?
- Allowing plugins to render under URLs other than `/a/...` in case they are specified as a standalone plugin page using the Grafana config (`[navigation.app_standalone_pages]`). Like this plugins cannot override or extend core functionality without explicit configuration done by the server admin.
- `isCorePage` - A new field in the `plugin.json` under `includes`. The name of this new field was just a quick idea, we can change it if we find a better one. It is needed because otherwise the backend doesn't know if it should render the page under the plugin section in case it was not configured as a standalone-plugin-page using `[navigation.app_standalone_pages]`. (Plugins can register page includes with custom URLs that can appear as "links" under the plugins nav section, which is fine, however if a plugin is registering a page include to extend / override core functionality, then it should only appear and work if it is configured correctly in Grafana as well.)
- I have updated the backend tests to use the .FindById() method instead of relying on indexes in the nav tree (I think it is more robust and less prone to errors in case the ordering would change in the future)
- I have tried to test the frontend changes out as much as possible, and they seem working - but I am still not sure I have covered all the edge cases.

### Example

**`plugin.json`**
```javascript
// ...
"includes": [
   // This page will show up under the app plugins section 
   {
      "type": "page",
      "name": "Random page",
      "path": "/a/my-random-app/random-page",
      "role": "Admin",
      "addToNav": true,
    },
   // With `isCorePage=true` this page will not show up under the plugins section, but can be configured 
   // to be used as a standalone page in the Grafana config 
    {
      "type": "page",
      "name": "Connect even more awesome data",
      "path": "/connections/connect-my-awesome-data",
      "role": "Admin",
      "addToNav": true,
      "isCorePage": true
    },
  ],
```


**`grafana.ini`**
```ini
# Specify a plugin page using its `path` and define which section it should live under
[navigation.app_standalone_pages]
"/connections/connect-my-awesome-data" = connections # section ID
```

leventebalogh committed:
feat(Routing): route standalone plugin pages to the `AppRoutePage`

torkelo commented on public/app/routes/routes.tsx:
```diff
@@ -540,3 +545,36 @@ export function getDynamicDashboardRoutes(cfg = config): RouteDescriptor[] {
+
+// Standalone plugin pages
+// These pages are registered by plugins and are showing up under different sections based on the Grafana configuration.
+function getStandalonePluginPageRoutes(): RouteDescriptor[] {
+  const state = store.getState();
+  const { navIndex } = state;
+  const isStandalonePage = (id: string) => id.includes('standalone-plugin-page-/');
+  const isDefaultAppRoute = (id: string) => id.includes('standalone-plugin-page-/a/'); // any page that is still served under "/a/:pluginId/*"
+  const standalonePageNavIds = Object.keys(navIndex).filter((id) => isStandalonePage(id) && !isDefaultAppRoute(id));
+
+  if (!standalonePageNavIds.length) {
+    return [];
+  }
+
+  return standalonePageNavIds.map((navId) => {
+    const { pluginId } = navIndex[navId];
+    const baseUrl = '';
+
+    return {
+      path: navIndex[navId].url || '',
+      exact: false,
+      component: ({ match, location, route, history, queryParams }) => (
+        <AppRootPage
+          route={route}
+          match={{ ...match, url: baseUrl, params: { ...match.params, pluginId } }}
+          queryParams={queryParams}
+          history={history}
+          location={location}
+        />
+      ),
+    };
+  });
```
This feels a bit messy. 

Feels like we might have an opportunity here to simply logic. What if we registered all app plugin routes this way, and insted of passing pluginId via match it was a regular prop. And we could also then pass in the navId, so we would no longer need the messy logic getPluginSection ?

mckn replied:
I really like this idea from @torkelo! So basically clean up the "regular" routes and use the `navIndex` in state as source to dynamically generate the routes needed by app plugins. If I understood everything correct? Then you would be able to pick up the pluginId from the state instead of passing it via the match and it can easily be passed to `AppRootPage` as a prop.

Do we have any scenarios where we could end up with apps missing from the state due to some reason and hence not being registered? I'm thinking pre-loading of app plugins etc.

torkelo replied:
yes, exactly. 

> Do we have any scenarios where we could end up with apps missing from the state due to some reason and hence not being registered? I'm thinking pre-loading of app plugins etc.

No all plugins are in the navIndex which is build on the backend based on plugin.json so does not need the plugin to be started.

leventebalogh replied:
Very good idea, thanks @torkelo and @mckn! 👍  

I have changed the code accordingly, please have a look if this is close to what you were thinking about. There is still a lot of opportunity for cleaning up and refactoring around the `AppRootPage`, but I would not go further than this in this PR.

 **A few notes on the new changes:**
-  We are now only allowing plugins to render under URLs other than `/a/...` if they are specified as a standalone plugin page using the Grafana config. Like this plugins cannot override or extend core functionality without explicit configuration done by the server admin.
- `isCorePage` - The name of this new field was just a quick idea, we can change it if we find a better one. It is needed because otherwise the backend doesn't know if it should render the page under the plugin section in case it was not configured as a standalone-plugin-page using `[navigation.app_standalone_pages]`. (Plugins can register page includes with custom URLs that can appear as "links" under the plugins nav section, which is fine, however if a plugin is registering a page include to extend / override core functionality, then it should only appear and work if it is configured correctly in Grafana as well.)

leventebalogh replied:
I have also updated the description of the PR to explain what is happening and to give an example on how to use this.

mckn reviewed (commented):
Great work on this! What @torkelo is suggesting sounds like a good approach forward.

mikkancso commented on public/app/routes/routes.tsx:
```diff
@@ -48,6 +50,9 @@ export function getAppRoutes(): RouteDescriptor[] {
     : [];
 
   return [
+    // Based on the Grafana configuration standalone plugin pages can even override and extend existing core pages, or they can register new routes under existing ones.
```
Please note that for this to work correctly, there has to be some config in Grafana configuration. This way Grafana admins know about these page overrides and have control over this feature. IMO that's a good enough safety mechanism.

mckn replied:
Sounds great!

leventebalogh committed:
chore: fix go lint issues

leventebalogh committed:
feat(plugin.json): introduce a new field called `isCorePage` for `includes`

leventebalogh committed:
chore: add explanatory comments for types

leventebalogh committed:
refactor(AppRootPage): receive the `pluginId` and `pluginSection` through the props

Now we are able to receive these as props as the pluginId is defined on navLinks
that are registered by plugins.

leventebalogh committed:
chore: update teests for AppRootPage

leventebalogh committed:
fix: remove rebase issue

leventebalogh committed:
tests(applinks): add a test for checking isCorePage plugin page setting

leventebalogh committed:
refactor(applinks): update tests to use FindById() and be more resilient to changes

leventebalogh committed:
fix: Go lint issues

mikkancso reviewed (approved):
LGTM!

mckn commented on packages/grafana-data/src/types/plugin.ts:
```diff
@@ -110,11 +110,18 @@ export interface PluginInclude {
+
+  // Marks a "page" type include as a route that can be used to extend or override existing Grafana pages.
+  // A page with this flag enabled will not show up under the plugins own sub-pages, but can be used as a standalone plugin page to show up under a certain navigation section using the `navigation.app_standalone_pages` Grafana setting.
+  isCorePage?: boolean;
```
Question: would it be possible to use the Grafana configuration to figure out if it can be used as a core page? I guess you need to specify something in the Grafana configuration to allow a plugin to override parts of the core pages?

mckn replied:
Basically do something like this and use the `plugin-id` from the configuration to figure this out instead of needing the plugin to specify that is should be served as a core page:

```bash
# Specify a plugin page using its `path` and define which section it should live under
[navigation.app_standalone_pages]
"/a/plugin-id/connections" = connections # section ID
```

leventebalogh replied:
Yeah, that's a very fair question and I was expecting it coming. :)

In most of the cases what you are suggesting should work. **The thing becomes tricky when we are trying to use a plugin to render under a core URL (not `/a/:plugin-id/...`).** If a plugin is exposing a page like that, then that page will probably only work and make sense if it is rendered under the correct section.

Example:

**`custom.ini`**
```ini
[navigation.app_standalone_pages]
"/connections/your-connections/integrations" = connections # section ID
```

**`plugin.json`**
```javascript
// ...
"includes": [
   {
      "type": "page",
      "name": "Random page",
      "path": "/a/my-random-app/random-page",
      "role": "Admin",
      "addToNav": true,
    },
   // The following page is expected to be rendered under `/connections/your-connections/integrations`, so it can
   // seamlessly extend Grafana's core capabilities. In case it wouldn't be configured in the INI, then it would show up
   // under the plugin's section and wouldn't work, as it is expected to live under a certain core URL. 
    {
      "type": "page",
      "name": "Integrations",
      "path": "/connections/your-connections/integrations",
      "role": "Admin",
      "addToNav": true,
      "isCorePage": true
    },
  ],
```

The edge-case is that currently we enable plugins to register custom paths (apart from `/a/:pluginId`) in their includes, and these will become links. In these cases though (if a page is meant to be extending core) they don't make sense as a simple link in case the server admin doesn't configure them.

Pages that live under `/a/:pluginId` can be more easily "moved around" in the nav-tree, as they will be always routed to the `AppRootPage` on the frontend and will always be displayed under their "own url". In these cases however these will be "just links" under other sections if they are used as standalone pages, and they won't give the feeling of an integrated feature.

Does this answer your question @mckn ?

mckn replied:
Yeah, I think I understand. A couple of follow up questions:

- Are you always replacing a whole section? Or do you replace e.g. one page of a section?
- Can we introduce Page IDs for the pages below a section that you are allowed to replace?

What about doing something like this (I have assumed that you want to replace single pages below a section):

`**plugin.json**`

```json
{ 
   "includes": [ 
      { 
         "type": "page", 
         "name": "Integrations", 
         "path": "/connections/your-connections/integrations", 
         "role": "Admin", 
         "addToNav": false, 
         "replaceInNav": {
            "section": "connections",
            "page": "integrations",
         }
      }, 
   ] 
}
```

Potentially added by the server admin:

```bash
# [navigation.app_sections.<Section ID>]
# Format: <Page ID> = <Plugin ID>
[navigation.app_sections.connections]
integrations = grafana-connections-app
```

I'm not sure this makes sense but if we only add a new attribute and don't reply on the `addToNav` then we won't risk of ending up in the edge case that you mentioned. Also - by using page ID we still keep the flexibility to refactor paths/routes within Grafana without breaking plugins. I guess a drawback with that is that we need to expose some kind of API in plugins to be able to resolve the path to a sectionId + pageId if you want to generate links in the plugin page.

I have probably missed something but would this work?

mckn replied:
And in the admin config you are basically saying: "This plugin is allowed to replace the navigation page in this section"

mckn replied:
So after rubber ducking this a bit. Ignore my comment above. What about the following:

1. Remove the `isCorePage` property.
2. Only try to add this behavior if the `addToNav` is `false`. 
3. Open an issue with the goal to improve the `[navigation.app_standalone_pages]` configuration to make it a bit more descriptive.

leventebalogh replied:
Created a follow up issue: https://github.com/grafana/grafana/issues/58336

mckn commented on public/app/routes/routes.tsx:
```diff
@@ -540,3 +531,25 @@ export function getDynamicDashboardRoutes(cfg = config): RouteDescriptor[] {
   ];
 }
+
+function getAppPluginRoutes(): RouteDescriptor[] {
+  const state = store.getState();
+  const { navIndex } = state;
+  const isStandalonePluginPage = (id: string) => id.startsWith('standalone-plugin-page-/');
+  const hasPluginIdSetForNav = (id: string) => navIndex[id]?.pluginId !== undefined;
+
+  return Object.keys(navIndex)
+    .filter(hasPluginIdSetForNav)
+    .map((navId) => {
+      const navItem = navIndex[navId];
+      const pluginNavSection = getRootSectionForNode(navItem);
+      const appPluginUrl = `/a/${navItem.pluginId}`;
+      const path = isStandalonePluginPage(navId) ? navItem.url || appPluginUrl : appPluginUrl; // Only standalone pages can use core URLs, otherwise we fall back to "/a/:pluginId"
+
+      return {
+        path,
+        exact: false, // route everything under this path to the plugin, so it can define more routes under this path
+        component: () => <AppRootPage pluginId={navItem.pluginId || ''} pluginNavSection={pluginNavSection} />,
+      };
+    });
+}
```
What do you think about doing something like this? Maybe also move this to its own file within the plugins folder. Similar to how we have done it with the plugin catalog routes.

```suggestion
function getAppPluginRoutes(): RouteDescriptor[] {
  const state = store.getState();
  const { navIndex } = state;
  const isStandalonePluginPage = (id: string) => id.startsWith('standalone-plugin-page-/');
  const isPluginNavModelItem = (model: NavModelItem): model is PluginNavModelItem =>
    'pluginId' in model && 'id' in model;

  return Object.values(navIndex)
    .filter<PluginNavModelItem>(isPluginNavModelItem)
    .map((navItem) => {
      const pluginNavSection = getRootSectionForNode(navItem);
      const appPluginUrl = `/a/${navItem.pluginId}`;
      const path = isStandalonePluginPage(navItem.id) ? navItem.url || appPluginUrl : appPluginUrl; // Only standalone pages can use core URLs, otherwise we fall back to "/a/:pluginId"

      return {
        path,
        exact: false, // route everything under this path to the plugin, so it can define more routes under this path
        component: () => <AppRootPage pluginId={navItem.pluginId} pluginNavSection={pluginNavSection} />,
      };
    });
}

interface PluginNavModelItem extends Omit<NavModelItem, 'pluginId' | 'id'> {
  pluginId: string;
  id: string;
}
```

mckn reviewed (commented):
Great work on this! Added some questions and a small nit.

leventebalogh committed:
refactor(routes): use cleaner types when working with plugin nav nodes

Co-authored-by: Marcus Andersson <marcus.andersson@grafana.com>

leventebalogh committed:
chore: fix linting issues

torkelo reviewed (approved):
Took this for a spin, and could not find any issues :) 

Think this looks good!

torkelo reviewed (commented):
The only thing I am not sure about is if we need isCorePage? Since it already needs to be configured in the ini file to allow the override?

leventebalogh commented:
Yeah, the only reason why I think we need it is because otherwise these pages would end up as a navigation items under the plugin's section in case they are not configured in the ini.

We can say though that we "don't care" about that edge-case, because that can also mean that a plugin is "not configured correctly".

I can imagine a scenario though where a plugin exposes a core page with a custom URL as an optional "feature", but also has other pages, and in this case it would make sense not to display the core pages unless they are configured in the ini.

leventebalogh commented:
We had a brainstorming session with @mckn now and we agreed that we remove `isCorePage` flag for now @torkelo. Instead of that people will be able to use `addToNav: false` in case they would like to hide a page from the plugin's navigation by default and only allow it to be registered as a standalone plugin page.

I will jump back on this Monday morning.

We were also talking about how we could make the config more flexible, hopefully we can work on that next week in a separate PR.

leventebalogh committed:
t: remove `isCorePage` field from includes
